scipy/sparse/linalg/dsolve/linsolve.py

```diff
@@ -383,3 +384,125 @@ def solve(b):
+    overwrite_b : bool, optional
+        Allow overwriting data in `b`.
+        This may give a performance gain.
+    Returns
+    -------
+    x : (M,) or (M, N) ndarray
+        Solution to the system `A x = b`.  Shape of return matches shape of `b`.
+    Raises
+    ------
+    LinAlgError
+        If `A` is singular.
+    ValueError
+        If shape of `A` or shape of `b` do not match the requirements.
+    Notes
+    -----
+    .. versionadded:: 0.20.0
+    """
+
+    ## check input
+    if A.ndim != 2:
+        raise ValueError('A must have 2 dimensions but its shape is {}.'.format(A.shape))
+    if A.shape[0] != A.shape[1]:
+        raise ValueError('A must be a square matrix but its shape is {}.'.format(A.shape))
+
+    if not isspmatrix_csr(A):
+        warn('CSR matrix format is required. Converting to CSR matrix.',
+             SparseEfficiencyWarning)
+        A = csr_matrix(A)
+
+    A.eliminate_zeros()
+    A.sort_indices()
+
+    b = np.asanyarray(b)
+
+    if b.ndim not in [1, 2]:
+        raise ValueError('b must have 1 or 2 dims but its shape is {}.'.format(b.shape))
+    if A.shape[0] != b.shape[0]:
+        raise ValueError('The size of the dimensions of A must be equal to '
+            'the size of the first dimension of b but the shape of A is '
+            '{} and the shape of b is {}.'.format(A.shape, b.shape))
+
+    ## init x
+    if not overwrite_b:
+        x = np.empty(b.shape)
+    else:
+        x = b
+
+    ## fill x
+    if lower:
+
+        ## fill forward
+        indptr_start = A.indptr[0]
+        for i in range(len(b)):
+            indptr_stop = A.indptr[i+1]
+
+            ## check regularity and triangularity
+            assert np.all(A.indices[indptr_start:indptr_stop-1] <
+                A.indices[indptr_start+1:indptr_stop])     # indices are assumed to be sorted
+            if indptr_stop <= indptr_start or A.indices[indptr_stop-1] < i:
+                raise LinAlgError('A is singular since the {}th diagonal '
+                    'entry of A is zero!'.format(i))
+            if A.indices[indptr_stop-1] > i:
+                raise LinAlgError('A is no lower triangle matrix, since the '
+                    'entry at ({},{}) is not zero!'.format(i, A.indices[indptr_stop-1]))
+
+            ## compute x[i]
+            x[i] = b[i]
+            if indptr_stop - indptr_start > 1:      # more than just the diagonal entry
+                A_column_indices_in_row_i = A.indices[indptr_start:indptr_stop-1]    # skip diagonal entry
+                A_values_in_row_i = A.data[indptr_start:indptr_stop-1]
+                if x.ndim > 1:
+                    A_values_in_row_i = A_values_in_row_i[:, np.newaxis]
+                x[i] -= np.sum(x[A_column_indices_in_row_i] * A_values_in_row_i, axis=0)
+            x[i] /= A.data[indptr_stop-1]           # devide by i-th diagonal element of A
+
+            ## next row
+            indptr_start = indptr_stop
+    else:
+
+        ## fill backward
```

Comment:
I don't like how this branch is essentially the same as the lower triangular case. Is it possible to move the lower/upper logic out and reuse the core `x[i]` computation for both cases?